src/D2L.CodeStyle.Analyzers/Helpers/ConstantAnalyzer.cs

```diff
@@ -0,0 +1,167 @@
+﻿using System.Collections.Immutable;
+using System.Linq;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Helpers {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class ConstantAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+			ImmutableArray.Create(
+				Diagnostics.NonConstantPassedToConstantParameter,
+				Diagnostics.InvalidConstantType
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis( GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics );
+			context.RegisterCompilationStartAction( CompilationStart );
+		}
+
+		public static void CompilationStart(
+			CompilationStartAnalysisContext context
+		) {
+			var constantAttribute = context.Compilation.GetTypeByMetadataName(
+				"D2L.CodeStyle.Annotations.Contract.ConstantAttribute"
+			);
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeParameters(
+					ctx,
+					(ParameterSyntax)ctx.Node,
+					constantAttribute
+				),
+				SyntaxKind.Parameter
+			);
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeArguments(
+					ctx,
+					(ArgumentSyntax)ctx.Node,
+					constantAttribute
+				),
+				SyntaxKind.Argument
+			);
+		}
+
+		private static void AnalyzeParameters(
+			SyntaxNodeAnalysisContext context,
+			ParameterSyntax parameterSyntax,
+			ISymbol constantAttribute
+		) {
+			var parameter = context.SemanticModel.GetDeclaredSymbol( parameterSyntax );
+
+			// Parameter is somehow null, so do nothing
+			if( parameter == null ) {
+				return;
+			}
+
+			// Parameter is not [Constant], so do nothing
+			if( !HasAttribute( parameter, constantAttribute ) ) {
+				return;
+			}
+
+			var type = parameter.Type;
+
+			// Special types (bool, enum, int, string, etc) are the only types
+			// which might be constant, aside from type parameters filled with
+			// special types
+			if( type.SpecialType != SpecialType.None ) {
+				return;
+			}

```

Comment:
Looking at `SpecialType` again, there's a number of them that aren't able to be constant.

```suggestion
			switch( type.SpecialType ) {
				case SpecialType.System_Enum:
				case SpecialType.System_Boolean:
				case SpecialType.System_Char:
				case SpecialType.System_SByte:
				case SpecialType.System_Byte:
				case SpecialType.System_Int16:
				case SpecialType.System_UInt16:
				case SpecialType.System_Int32:
				case SpecialType.System_UInt32:
				case SpecialType.System_Int64:
				case SpecialType.System_UInt64:
				case SpecialType.System_Decimal:
				case SpecialType.System_Single:
				case SpecialType.System_Double:
				case SpecialType.System_String:
					return;
				default:
					break;
			}
```